tests/integration/test_prometheus_scrape_multiunit.py

```diff
@@ -0,0 +1,188 @@
+    await asyncio.gather(
+        *[check_prometheus_is_ready(ops_test, prometheus_app_name, u) for u in range(num_units)]
+    )
+
+    # WHEN prometheus is not related to anything
+    # THEN all prometheus units should have only one scrape target (self-scraping)
+    for unit_num in range(num_units):
+        targets = await get_prometheus_active_targets(ops_test, prometheus_app_name, unit_num)
+        assert len(targets) == 1
+        self_scrape = next(iter(targets))
+        assert self_scrape["labels"]["job"] == "prometheus"
+        assert self_scrape["labels"]["instance"] == "localhost:9090"
+
+    # WHEN prometheus is related to the tester
+    await ops_test.model.add_relation(prometheus_app_name, tester_app_name)
+    await ops_test.model.wait_for_idle(apps=app_names, status="active")
+
+    # THEN all prometheus units should have all tester units as targets (as well as self-scraping)
+    # `targets_by_unit` is a List[List[dict]]: every unit has a List[dict] targets.
+    targets_by_unit = await asyncio.gather(
+        *[
+            get_prometheus_active_targets(ops_test, prometheus_app_name, u)
+            for u in range(num_units)
+        ]
+    )
+    assert all(len(targets) == num_units + 1 for targets in targets_by_unit)
+
+    # AND all prometheus units have the exact same targets
+    # Only comparing the `labels` because comparing the entire `targets` dict would be cumbersome:
+    # would need to pop 'lastScrape', 'lastScrapeDuration', whose values may differ across units.
+    labels = [[{"labels": d["labels"]} for d in unit_targets] for unit_targets in targets_by_unit]
+    for u in range(1, len(targets_by_unit)):
+        assert DeepDiff(labels[0], labels[u], ignore_order=True) == {}
+    # Could use `set`, but that would produce unhelpful error messages.
+    # assert len(set(map(lambda x: json.dumps(x, sort_keys=True), targets_by_unit))) == 1
+
+
+@pytest.mark.abort_on_fail
+async def test_upgrade_prometheus(ops_test: OpsTest, prometheus_charm):
+    """Upgrade prometheus and confirm all is still green (see also test_upgrade_charm.py)."""
+    # WHEN prometheus is upgraded
+    await ops_test.model.applications[prometheus_app_name].refresh(
+        path=prometheus_charm, resources=prometheus_resources
+    )
+
+    # THEN nothing breaks
+    await ops_test.model.wait_for_idle(status="active", idle_period=60)
+    await asyncio.gather(
+        *[check_prometheus_is_ready(ops_test, prometheus_app_name, u) for u in range(num_units)]
+    )
+
+
+@pytest.mark.abort_on_fail
+async def test_rescale_prometheus(ops_test: OpsTest):
+    # WHEN prometheus is scaled up
+    num_additional_units = 1
+    await ops_test.model.applications[prometheus_app_name].scale(scale_change=num_additional_units)
+    new_num_units = num_units + num_additional_units
+
+    # THEN nothing breaks
+    await ops_test.model.wait_for_idle(
+        apps=[prometheus_app_name],
+        status="active",
+        timeout=600,
```

Comment:
This timeout is probably really excessive. Even 300 is a lot better, but if we can't scale in 60 seconds or less, even on GH, something is dramatically wrong.

Reply:
Done in next commit.